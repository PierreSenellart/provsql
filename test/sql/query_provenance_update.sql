\set ECHO none
\pset format unaligned
SET search_path TO provsql_test,provsql;

CREATE TABLE update_test(id INT PRIMARY KEY);

INSERT INTO update_test (id)
VALUES (1), (2), (3), (4), (5);

SELECT add_provenance('update_test');

DELETE FROM query_provenance;

-- Test 1: single row update
SELECT create_provenance_mapping('update_test_id', 'update_test', 'id');
SELECT COUNT(*) FROM update_test;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 1;

DROP TABLE update_test_id;
UPDATE FROM update_test SET id = 6 WHERE id = 1;
SELECT create_provenance_mapping('update_test_id', 'update_test', 'id');
SELECT COUNT(*) FROM update_test_id;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 1;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 6;

-- Test 2: update of updated row (old row and new row)
DROP TABLE update_test_id;
UPDATE FROM update_test SET id = 7 WHERE id = 1;
UPDATE FROM update_test SET id = 8 WHERE id = 6;
SELECT create_provenance_mapping('update_test_id', 'update_test', 'id');
SELECT COUNT(*) FROM update_test_id;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 1;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 6;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 7;
SELECT probability_evaluate(provenance) FROM update_test_id WHERE value = 8;

-- Test 3: multiple rows deletion
DROP TABLE update_test_id;
UPDATE FROM update_test SET id = id + 10 WHERE id >= 2 AND id <= 4;
SELECT create_provenance_mapping('update_test_id', 'update_test', 'id');
SELECT COUNT(*) FROM update_test_id;
SELECT value, probability_evaluate(provenance) FROM update_test_id ORDER BY value;

-- Test 4: update token tracking/ logging into query_provenance table
SELECT query FROM query_provenance ORDER BY ts;

DELETE FROM query_provenance;

DROP TABLE delete_test;
DROP TABLE delete_test_id;
