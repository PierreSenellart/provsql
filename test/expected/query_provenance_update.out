\set ECHO none
add_provenance

(1 row)
create_provenance_mapping

(1 row)
count
5
(1 row)
probability_evaluate
1
(1 row)
create_provenance_mapping

(1 row)
count
5
(1 row)
probability_evaluate
0
(1 row)
probability_evaluate
1
(1 row)
create_provenance_mapping

(1 row)
count
5
(1 row)
probability_evaluate
0
(1 row)
probability_evaluate
0
(1 row)
probability_evaluate
0
(1 row)
probability_evaluate
1
(1 row)
create_provenance_mapping

(1 row)
count
5
(1 row)
value|probability_evaluate
1|0
2|0
3|0
4|0
5|1
6|0
7|0
8|1
12|1
13|1
14|1
(11 rows)
query
UPDATE FROM update_test SET id = 6 WHERE id = 1;
UPDATE FROM update_test SET id = 7 WHERE id = 1;
UPDATE FROM update_test SET id = 8 WHERE id = 6;
UPDATE FROM update_test SET id = id + 10 WHERE id >= 2 AND id <= 4;
(4 rows)
